core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -957,10 +991,28 @@ class TimeSeriesShard(val dataset: Dataset,
     true
   }
 
-  // Permanently removes the given partition ID from our in-memory data structures
-  // Also frees partition key if necessary
+  /**
+   * Permanently removes the given partition ID from our in-memory data structures
+   * Also frees partition key if necessary
+   * Caller must _not_ hold partSetLock write lock.
+   */
   private def removePartition(partitionObj: TimeSeriesPartition): Unit = {
-    partSet.remove(partitionObj)
+    doRemovePartition(partitionObj, false)
+  }
+
+  /**
+   * Permanently removes the given partition ID from our in-memory data structures
+   * Also frees partition key if necessary
+   * @param locked true if caller holds partSetLock write lock
+   */
+  private def doRemovePartition(partitionObj: TimeSeriesPartition, locked: Boolean): Unit = {
+    if (locked) {
+      partSet.remove(partitionObj)
+    } else {
+      val stamp = partSetLock.writeLock()
```

Comment:
This statement is in the ingestion path. We should make sure this will exit and not "hang" or "block" ingestion.

Reply:
Do you mean an exception? If an exception is thrown from this method, then the underlying data structures used here are corrupt. Ensuring the lock is released doesn't fix that.

Reply:
No. What I meant is that writeLock method call is a blocking call which wont return until lock is obtained. Did we evaluate the possibility of it hanging for a long time? Should there be a timeout ?

Reply:
I was referring to https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/StampedLock.html#writeLock--

> public long writeLock()
Exclusively acquires the lock, blocking if necessary until available.

Reply:
Other than ingestion, the query acquires a read lock when calling getPartition. It should only be holding the lock on the order of a microsecond at most. No significant blocking should occur on the ingestion path.